danger/src/PullRequestManager.ts

```diff
@@ -0,0 +1,165 @@
+
+export enum ChangelogEntryType {
+  BUG_FIXES = 'bug-fix',
+  NEW_FEATURES = 'new-feature',
+  BREAKING_CHANGES = 'breaking-change',
+}
+
+export const DEFAULT_ENTRY_TYPE = ChangelogEntryType.BUG_FIXES;
+
+export type ChangelogEntry = {
+  type: ChangelogEntryType;
+  message: string;
+};
+
+export const DEFAULT_CHANGELOG_ENTRY_KEY = 'default' as const;
+
+export type ChangelogEntries = {
+  [DEFAULT_CHANGELOG_ENTRY_KEY]: ChangelogEntry;
+  [key: string]: ChangelogEntry;
+};
+
+export type PullRequest = GitHubPRDSL | Octokit.PullsListResponseItem;
+
+const dangerMessage = `Add missing changelog`;
+const dangerTags = `[danger][bot]`;
+
+export class PullRequestManager {
+  constructor(private pullRequest: PullRequest, private githubApi: GithubApiWrapper) {}
+
+  /**
+   * Get suggested changelog entries from PR provided in the constructor.
+   *
+   * If PR doesn't contais `# Changelog` section, this method returns:
+   * {
+   *   [DEFAULT_CHANGELOG_ENTRY_KEY]: <title of this pr without tags>
+   * }
+   * Otherwise, it tries to parse PR's body.
+   */
+  parseChangelogSuggestionFromDescription(): ChangelogEntries {
+    const changelogEntries = {
+      [DEFAULT_CHANGELOG_ENTRY_KEY]: {
+        type: DEFAULT_ENTRY_TYPE,
+        message: this.pullRequest.title.replace(/\[.*\]/, '').trim(),
+      },
+    };
+
+    const changelogTag = this.pullRequest.body
+      .match(/#\schangelog(([^#]*?)\s?)*/i)?.[0]
+      ?.replace(/^-/, '');
+    if (changelogTag) {
+      changelogTag
+        .split('\n')
+        .slice(1)
+        .map(line => line.trim())
+        .filter(line => line.length > 0)
+        .forEach(line => {
+          const tags = this.parseTagsFromLine(line);
+          if (!tags) {
+            warn(`Couldn't parse line: ${line}.`);
+            return;
+          }
+
+          changelogEntries[tags.packageName] = {
+            type: tags.type,
+            message: line.replace(/\[.*\]/, '').trim(),
+          };
+        });
+    }
+
+    return changelogEntries;
+  }
+
+  async createOrUpdateRPAsync(
+    missingEntries: { packageName: string; content: string }[]
+  ): Promise<PullRequest | null> {
+    const dangerHeadRef = `@danger/add-missing-changelog-to-${this.pullRequest.number}`;
+    const dangerBaseRef = this.pullRequest.head.ref;
+
+    const fileMap = missingEntries.reduce(
+      (prev, current) => ({
+        ...prev,
+        [getPackageChangelogRelativePath(current.packageName)]: current.content,
+      }),
+      {}
+    );
+
+    await this.githubApi.createOrUpdateBranchFromFileMap(fileMap, {
+      baseBranchName: dangerBaseRef,
+      branchName: dangerHeadRef,
+      message: `${dangerTags} ${dangerMessage}`,
+    });
+
+    const prs = await this.githubApi.getOpenPRs({
+      fromBranch: dangerHeadRef,
+      toBranch: dangerBaseRef,
+    });
+
+    if (prs.length > 1) {
+      warn("Couldn't find the correct pull request. Too many open ones.");
```

Comment:
Is this even possible to have two PRs opened with the same branches? 😮 I'm just curious 😄

Reply:
That's a good question 🤔And it's not possible 😅

Reply:
However, I leave this check if something goes very bad ;)